fix record batch writing to dataset

TevinWang opened the pull request:
Implement writing dataset with `Iterator[RecordBatch]`, like mentioned in the docs. This requires an argument of schema, so we throw a ValueError if it doesn't exist.

Minor ugliness: I check for `Iterator` type, not `Iterator[RecordBatch]`. Let me know if this causes some concerns.

Closes #1019

changhiskhan commented on python/python/lance/dataset.py:
```diff
@@ -942,6 +945,13 @@ def _coerce_reader(
+    elif isinstance(data_obj, Iterator):
+        if schema is not None:
+            return pa.RecordBatchReader.from_batches(schema, data_obj)
+        else:
```
Can we use the schema from the first batch?

TevinWang replied:
From what I recall, we can't access that batch anymore if we expose it in the iterator right? Let me know if I'm misunderstanding.

wjones127 replied:
Yeah we'd have to implement a peekable iterator. I don't think we need to do that for now. I think beginners will simply write tables, and advanced users won't have much issue providing the schema themselves.

changhiskhan commented on python/python/lance/dataset.py:
```diff
@@ -942,6 +945,13 @@ def _coerce_reader(
         return pa.dataset.Scanner.from_dataset(data_obj).to_reader()
     elif isinstance(data_obj, pa.dataset.Scanner):
         return data_obj.to_reader()
+    elif isinstance(data_obj, Iterator):
```
add unit test please

changhiskhan reviewed (commented):
Also what about an Iterable/Sequence (e.g., list of RecordBatch)

Also worth improving the error message here: https://github.com/lancedb/lance/pull/1030/files#diff-3c0e591b6a64801b9e3f92380ab99c1546e03faaa769b6de4b37af58d95c588cR958 

It just says "Uknown data_obj type" which as you can see from #1019 it's pretty cryptic

changhiskhan reviewed (approved):
Thanks @TevinWang

TevinWang commented:
@changhiskhan rerequesting review here, I changed some of the logic to include iterables, but needed to move the conditionals since a RecordBatchReader is an iterable (also added unit tests). Let me know if there are any more changes.

changhiskhan reviewed (approved):
the new unit tests are almost identical. please collapse this into a single method
